### PR TITLE
feature: Added documentation with sphinx

### LINKED ISSUE #270: Add sphinx support for local html documentation files
#219

### PR DESCRIPTION
**Pull request by dkiechle**
Fixed #270 and #271 

My contribution is licensed under the MIT license.

### CODE REVIEW

**Comment by SebastianNiemann** [comment]
I found a critical bug:

```
.. cpp:function:: template <T, N1, N2, N3> constexpr polynomials(parameters)

  .. versionadded:: 1.x
  
  **Template parameters**
   
    * **T** - The value type within the parameters and polynomials.
    * **N1** (``std::size_t``) - The number of indeterminates.
    * **N2** (``std::size_t``) - The largest degree.
    * **N3** (``std::size_t``) - The number of parameters.

  **Function parameters**
   
    * **parameter** (``std::array<T, N1*N3>``) - The parameters.

  **Returns**
   
    ``std::array<T, number_of_coefficients(N1, N2)>`` - The polynomials.
  
  Returns a polynomial with ``N1`` indeterminates and largest degree ``N2`` for each of the ``N3`` parameters (all with length ``N1``) in ``parameters``, written as tuple with one polynomial term per element.
  
  The indeterminates :math:`(x_1, \ldots, x_\text{N1})` of the ``n``-th polynomial are substituted by ``(parameters[n], ..., parameters[n + N1-1])`` and all coefficients are ``1.0``.
  
  The result is filled by one polynomial after another, going from the first parameter's polynomial to the last. The terms within each tuple are going from the terms with degree ``N2`` (in lexicographic order of the indeterminates indices) to degree 0. The last value (degree 0) will always be ``1.0``.
  
  Will return ``{1.0, ..., 1.0}`` (length ``N3``) if either ``N1`` or ``N2`` is ``0``. 
  
  .. code-block:: c++
  
    #include <mantella0>
    
    int main() {
      constexpr std::size_t number_of_indeterminates = 2;
      constexpr std::size_t largest_degree = 2;
      constexpr std::size_t number_of_parameters = 3;
      
      constexpr std::array<double, number_of_indeterminates> parameters = {0.2, -9.0, 0.1, 3.0};
      constexpr auto polynomials = mant::polynomials<number_of_indeterminates, largest_degree, number_of_parameters>(parameters);
      
      auto begin_parameters = parameters.cbegin();
      auto begin_polynomials = polynomials.cbegin();
      const auto serialise_array = [](const std::string& serialised_array, const auto element) { return serialised_array + ", " + element; };
      for (std::size_t n = 0; n < number_of_parameters; ++n) {
        std::cout << std::accumulate(begin_parameters, std::next(begin_parameters + number_of_indeterminates, "", serialise_array) << " -> "<< std::accumulate(begin_polynomials, std::next(begin_polynomials + number_of_indeterminates, "", serialise_array) << std::endl;
        
        std::advance(begin_parameters, number_of_indeterminates);
        std::advance(begin_polynomials, number_of_indeterminates);
      }
      
      return 0;
    }
```

Will be missing the first dot of `.. code-block:: c++` in the `rst`-File and therefore not be rendered correctly.

**Comment by SebastianNiemann** [comment]
After fixing the aforementioned bug, I have now an even more serious bug.

```
#pragma once

/**
.. cpp:function:: template <T, N1, N2, N3> constexpr polynomials(parameters)

  .. versionadded:: 1.x
  
  **Template parameters**
   
    * **T** - The value type within the parameters and polynomials.
    * **N1** (``std::size_t``) - The number of indeterminates.
    * **N2** (``std::size_t``) - The largest degree.
    * **N3** (``std::size_t``) - The number of parameters.

  **Function parameters**
   
    * **parameters** (``std::array<T, N1*N3>``) - The parameters.

  **Returns**
   
    ``std::array<T, N3*number_of_coefficients(N1, N2)>`` - The polynomials.
  
  Returns a polynomial with ``N1`` indeterminates and largest degree ``N2`` for each of the ``N3`` parameters (all with length ``N1``) in ``parameters``, written as tuple with one polynomial term per element.
  
  The indeterminates :math:`(x_1, \ldots, x_\text{N1})` of the ``n``-th polynomial are substituted by ``(parameters[n], ..., parameters[n + N1-1])`` and all coefficients are ``1.0``.
  
  The result is filled by one polynomial after another, going from the first parameter's polynomial to the last. The terms within each tuple are going from the terms with degree ``N2`` (in lexicographic order of the indeterminates indices) to degree 0. The last value (degree 0) will always be ``1.0``.
  
  Will return ``{1.0, ..., 1.0}`` (length ``N3``) if either ``N1`` or ``N2`` is ``0``. 
  
  .. code-block:: c++
  
    #include <mantella0>
    
    int main() {
      constexpr std::size_t number_of_indeterminates = 2;
      constexpr std::size_t largest_degree = 2;
      constexpr std::size_t number_of_parameters = 3;
      
      constexpr std::array<double, number_of_indeterminates> parameters = {0.2, -9.0, 0.1, 3.0};
      constexpr auto polynomials = mant::polynomials<number_of_indeterminates, largest_degree, number_of_parameters>(parameters);
      
      auto begin_parameters = parameters.cbegin();
      auto begin_polynomials = polynomials.cbegin();
      const auto serialise_array = [](const std::string& serialised_array, const auto element) { return serialised_array + ", " + element; };
      for (std::size_t n = 0; n < number_of_parameters; ++n) {
        std::cout << std::accumulate(begin_parameters, std::next(begin_parameters + number_of_indeterminates, "", serialise_array) << " -> "<< std::accumulate(begin_polynomials, std::next(begin_polynomials + number_of_indeterminates, "", serialise_array) << "\n";
        
        std::advance(begin_parameters, number_of_indeterminates);
        std::advance(begin_polynomials, number_of_indeterminates);
      }
      std::cout << std::endl;
      
      return 0;
    }
*/
template <typename T, std::size_t N1, std::size_t N2, std::size_t N3>
constexpr std::array<T, number_of_coefficients(N1, N2)> polynomials(
    const std::array<T, N1*N3>& parameters);
    
/**
.. cpp:function:: template <T, N1, N2> constexpr polynomial(parameter)

  .. versionadded:: 1.x
  
  **Template parameters**
   
    * **T** - The value type within the parameter and polynomial.
    * **N1** (``std::size_t``) - The number of indeterminates.
    * **N2** (``std::size_t``) - The largest degree.

  **Function parameters**
   
    * **parameter** (``std::array<T, N1>``) - The parameter.

  **Returns**
   
    ``std::array<T, number_of_coefficients(N1, N2)>`` - The polynomial.
  
  Short-cut for ``polynomials<T, N1, N2, 1>(parameter)``.
*/
template <typename T, std::size_t N1, std::size_t N2>
constexpr std::array<T, number_of_coefficients(N1, N2)> polynomial(
    const std::array<T, N1>& parameters);
    
```

Gets converted into
```

.. cpp:function:: template <T, N1, N2, N3> constexpr polynomials(parameters)

  .. versionadded:: 1.x
  
  **Template parameters**
   
    * **T** - The value type within the parameters and polynomials.
    * **N1** (``std::size_t``) - The number of indeterminates.
    * **N2** (``std::size_t``) - The largest degree.
    * **N3** (``std::size_t``) - The number of parameters.

  **Function parameters**
   
    * **parameters** (``std::array<T, N1*N3>``) - The parameters.

  **Returns**
   
    ``std::array<T, N3*number_of_coefficients(N1, N2)>`` - The polynomials.
  
  Returns a polynomial with ``N1`` indeterminates and largest degree ``N2`` for each of the ``N3`` parameters (all with length ``N1``) in ``parameters``, written as tuple with one polynomial term per element.
  
  The indeterminates :math:`(x_1, \ldots, x_\text{N1})` of the ``n``-th polynomial are substituted by ``(parameters[n], ..., parameters[n + N1-1])`` and all coefficients are ``1.0``.
  
  The result is filled by one polynomial after another, going from the first parameter's polynomial to the last. The terms within each tuple are going from the terms with degree ``N2`` (in lexicographic order of the indeterminates indices) to degree 0. The last value (degree 0) will always be ``1.0``.
  
  Will return ``{1.0, ..., 1.0}`` (length ``N3``) if either ``N1`` or ``N2`` is ``0``. 
  
  .. code-block:: c++
  
    #include <mantella0>
    
    int main() {
      constexpr std::size_t number_of_indeterminates = 2;
      constexpr std::size_t largest_degree = 2;
      constexpr std::size_t number_of_parameters = 3;
      
      constexpr std::array<double, number_of_indeterminates> parameters = {0.2, -9.0, 0.1, 3.0};
      constexpr auto polynomials = mant::polynomials<number_of_indeterminates, largest_degree, number_of_parameters>(parameters);
      
      auto begin_parameters = parameters.cbegin();
      auto begin_polynomials = polynomials.cbegin();
      const auto serialise_array = [](const std::string& serialised_array, const auto element) { return serialised_array + ", " + element; };
      for (std::size_t n = 0; n < number_of_parameters; ++n) {
        std::cout << std::accumulate(begin_parameters, std::next(begin_parameters + number_of_indeterminates, "", serialise_array) << " -> "<< std::accumulate(begin_polynomials, std::next(begin_polynomials + number_of_indeterminates, "", serialise_array) << "\n";
        
        std::advance(begin_parameters, number_of_indeterminates);
        std::advance(begin_polynomials, number_of_indeterminates);
      }
      std::cout << std::endl;
      
      return 0;
    }

.. cpp:function:: template <T, N1, N2> constexpr polynomial(parameter)
```

See that anything after `.. cpp:function:: template <T, N1, N2> constexpr polynomial(parameter)` is missing.

**Comment by SebastianNiemann** [comment]
In addition
```
#pragma once

/**
.. cpp:function:: template <N1, N2> constexpr number_of_coefficients()

  .. versionadded:: 1.x
  
  **Template parameters**
   
    * **N1** (``std::size_t``) - The number of indeterminates.
    * **N2** (``std::size_t``) - The largest degree.

  **Returns**
   
    ``std::size_t`` - The number of coefficients.
  
  Calculates the number of coefficients in a polynomial with ``N1`` indeterminates and largest degree ``N2``.
    
  The number of coefficients will be equal to:

  .. math::
    
    1 + \sum_{d = 1}^\text{N2} \binom{\text{N1}}{d}
  
  In consequence, it will return ``1`` if either ``N1`` or ``N2`` is ``0``. 
  
  This is typically used together with :cpp:func:`polynomial`, to calculate the number of coefficients to be stored the resulting ``std::array`` at compile-time.
*/
template <std::size_t N1, std::size_t N2>
constexpr std::size_t number_of_coefficients;
```

Gets converted into an empty file.

**Comment by dkiechle** [comment]
Finished issues #270 and #271

**Comment by SebastianNiemann** [comment]
@dkiechle As this is an old pull request and I want to be sure: Is my assumption correct, that this is ready to be included from your point of view?

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,59 @@ Makefile
 cmake_install.cmake
 install_manifest.txt
```
Comment (SebastianNiemann) [inline .gitignore]: This file was not actually merged, as the conflict markers all still present

File: doc/.prepare_doc.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+an_error_occured=False
+
+hppfiles = []
+for path, subdirs, files in os.walk('../include/mantella0_bits/'):
+  for file in files:
+    if file.endswith('.hpp'):
+       hppfiles = hppfiles + [(os.path.join(path, file), os.path.join(path.replace('../include/mantella0_bits', './api_reference', 1), file.split('.')[0] + '.rst'))]
+
+# file[0] = source file; file[1] = destination file
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: What is the meaning behind this comment?

File: doc/.prepare_doc.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+          continue
+          
+        # Write result from code file in .rst
+        docfile.write('\n' + part[2] + '.. code-block:: none\n')
+        for line in output.stdout.read().split('\n'):
+          docfile.write('\n' + part[2] + '  ' + line)
+        docfile.write('\n')
+          
+      if 'image' in part[3]:
+
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: Delete empty line

File: doc/.prepare_doc.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        output.wait()
+        if output.returncode != 0:
+          an_error_occured=True
+          print bcolors.FAIL + '  Failure during genarate image: \n' + bcolors.ENDC + output.stderr.read()
+          continue
+        
+        docfile.write('\n' + part[2] + '.. image:: ../assert/images/' + part[4] + '\n')
+      
+    docfile.close()
+    print '\x1b[2K \r',
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: Magic command

File: doc/.prepare_doc.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+if os.path.exists('./tmp'): shutil.rmtree('./tmp')
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: No oneliner for if-conditions please.

File: doc/_static/css/landingpage.css [hunk 6]
```diff
@@ -0,0 +1,1526 @@
+html {
```
Comment (SebastianNiemann) [inline doc/_static/css/landingpage.css]: As this file is named `doc/_static/css/landingpage.css`, it shouldn't be part of this pull request.

File: doc/conf.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+# Project information
+project = u'Mantella'
+copyright = u'2016, Sebastian Niemann, Oskar Pusz, Daniel Kiechle, Wolf Bock von Wülfingen and Philipp Schiffmann'
+author = u'Mantella team and contributors'
+github_url = 'https://github.com/SebastianNiemann/Mantella'
+
+version = u'0.50'
```
Comment (SebastianNiemann) [inline doc/conf.py]: 0.61

File: doc/conf.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+release = u'0.50.0'
```
Comment (SebastianNiemann) [inline doc/conf.py]: 0.61.0

Can't this be build from `version` with the patch version (`.0`) added to it?

File: doc/conf.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'std'),
+  ('cpp:type', 'std::size_t'),
+  ('cpp:type', 'std::array'),
+  ('cpp:type', 'std::vector'),
+  ('cpp:type', 'std::function'),
+  ('cpp:type', 'std::pair'),
+  ('cpp:type', 'std::string'),
+  ('cpp:type', 'optimisation_problem'),
+  ('cpp:type', 'begin_parameter'),
+  ('cpp:type', 'end_parameter'),
```
Comment (SebastianNiemann) [inline doc/conf.py]: Does not exists in Mantella anymore

File: doc/conf.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'N'),
+  ('cpp:type', 'std'),
+  ('cpp:type', 'std::size_t'),
+  ('cpp:type', 'std::array'),
+  ('cpp:type', 'std::vector'),
+  ('cpp:type', 'std::function'),
+  ('cpp:type', 'std::pair'),
+  ('cpp:type', 'std::string'),
+  ('cpp:type', 'optimisation_problem'),
+  ('cpp:type', 'begin_parameter'),
```
Comment (SebastianNiemann) [inline doc/conf.py]: Does not exists in Mantella anymore

File: doc/conf.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'T2'),
+  ('cpp:type', 'N'),
+  ('cpp:type', 'std'),
+  ('cpp:type', 'std::size_t'),
+  ('cpp:type', 'std::array'),
+  ('cpp:type', 'std::vector'),
+  ('cpp:type', 'std::function'),
+  ('cpp:type', 'std::pair'),
+  ('cpp:type', 'std::string'),
+  ('cpp:type', 'optimisation_problem'),
```
Comment (SebastianNiemann) [inline doc/conf.py]: Does not exists in Mantella anymore

File: doc/conf.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'T1'),
+  ('cpp:type', 'T2'),
+  ('cpp:type', 'N'),
+  ('cpp:type', 'std'),
+  ('cpp:type', 'std::size_t'),
+  ('cpp:type', 'std::array'),
+  ('cpp:type', 'std::vector'),
+  ('cpp:type', 'std::function'),
+  ('cpp:type', 'std::pair'),
+  ('cpp:type', 'std::string'),
```
Comment (SebastianNiemann) [inline doc/conf.py]: Does not exists in Mantella anymore

File: doc/conf.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'T'),
+  ('cpp:type', 'T1'),
+  ('cpp:type', 'T2'),
+  ('cpp:type', 'N'),
+  ('cpp:type', 'std'),
+  ('cpp:type', 'std::size_t'),
+  ('cpp:type', 'std::array'),
+  ('cpp:type', 'std::vector'),
+  ('cpp:type', 'std::function'),
+  ('cpp:type', 'std::pair'),
```
Comment (SebastianNiemann) [inline doc/conf.py]: Does not exists in Mantella anymore

File: doc/conf.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+  ('cpp:type', 'std::string'),
+  ('cpp:type', 'optimisation_problem'),
+  ('cpp:type', 'begin_parameter'),
+  ('cpp:type', 'end_parameter'),
+  ('cpp:type', 'parameter'),
+]
+
+master_doc = 'mantella'
+source_suffix = ['.rst', '.md']
```
Comment (SebastianNiemann) [inline doc/conf.py]: Do we still need `.rst` as source file format?

File: doc/_static/css/landingpage.css (hunk 6 above)
Comment (SebastianNiemann) [inline doc/_static/css/landingpage.css]: The same holds true for some other images (not all, I think the image containing only the frog logo was part of the api documentation), fonts and misc. files.

**Comment by OpusV** [comment]
Fixed #270 #271

File: doc/mantella.rst [hunk 15]
```diff
@@ -0,0 +1,15 @@
+Welcome to Mantella's documentation!
+====================================
+
+Contents:
+
+.. toctree::
+  :maxdepth: 3
+
+
```
Comment (SebastianNiemann) [inline doc/mantella.rst]: Why 3 newlines?

File: doc/.prepare_doc.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+          continue
+        
+        docfile.write('\n' + part[2] + '.. image:: ../assert/images/' + part[4] + '\n')
+      
+    docfile.close()
+    print '\x1b[2K \r', #clear line
+
+if os.path.exists('./tmp'):
+  shutil.rmtree('./tmp')
+if an_error_occured:
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: Please add a newline before this if-condition

File: doc/.prepare_doc.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+  # Continues with the next file if no comments where found
+  if not comments: 
+    print bcolors.FAIL + '  No comments found' + bcolors.ENDC
+    sys.stdout.flush()
+    continue
+
+  if not os.path.exists(os.path.dirname(file[1])):
+    os.makedirs(os.path.dirname(file[1]))
+      
+  #TODO Add comments to the regex below and add a large comment block describing the structure returned by `re.findall` (this will be done by Sebastian)
```
Comment (SebastianNiemann) [inline doc/.prepare_doc.py]: Is this still to be done?
